functions/create_welcome_message.ts

```diff
@@ -28,34 +30,104 @@ export const WelcomeMessageSetupFunction = DefineFunction({
       },
     },
-    required: ["welcome_message", "channel"],
+    required: ["message", "channel"],
   },
 });
 
 export default SlackFunction(
   WelcomeMessageSetupFunction,
   async ({ inputs, client }) => {
+    const { channel, message, author } = inputs;
     const uuid = crypto.randomUUID();
 
-    const putResponse = await client.apps.datastore.put({
-      datastore: DATASTORE_NAME,
-      item: {
-        id: uuid,
-        channel: inputs.channel,
-        message: inputs.welcome_message,
-        author: inputs.author,
-      },
+    // Save information about the welcome message to the datastore
+    const putResponse = await client.apps.datastore.put<
+      typeof WelcomeMessageDatastore.definition
+    >({
+      datastore: WelcomeMessageDatastore.name,
+      item: { id: uuid, channel, message, author },
     });
 
     if (!putResponse.ok) {
-      return {
-        error: putResponse.error,
-        outputs: {},
-      };
-    } else {
-      return {
-        outputs: {},
-      };
+      return { error: `Failed to save welcome message: ${putResponse.error}` };
     }
+
+    // Search for any existing triggers for the welcome workflow
+    const triggers = await findUserJoinedChannelTrigger(client, channel);
+    if (triggers.error) {
+      return { error: `Failed to lookup existing triggers: ${triggers.error}` };
+    }
+
+    // Create a new user_joined_channel trigger if none exist
+    if (!triggers.exists) {
+      const newTrigger = await saveUserJoinedChannelTrigger(client, channel);
+      if (!newTrigger.ok) {
+        return {
+          error: `Failed to create welcome trigger: ${newTrigger.error}`,
+        };
+      }
+    }
+
+    return { outputs: {} };
   },
 );
+
+/**
+ * findUserJoinedChannelTrigger returns if the user_joined_channel trigger
+ * exists for the "Send Welcome Message" workflow in a channel.
+ */
+export async function findUserJoinedChannelTrigger(
+  client: SlackAPIClient,
+  channel: string,
+): Promise<{ error?: string; exists?: boolean }> {
+  // Collect all existing triggers created by the app
+  const allTriggers = await client.workflows.triggers.list({ is_owner: true });
+  if (!allTriggers.ok) {
+    return { error: allTriggers.error };
+  }
+
+  // Find user_joined_channel triggers for the "Send Welcome Message"
+  // workflow in the specified channel
+  const joinedTriggers = allTriggers.triggers.filter((trigger) => (
+    trigger.workflow.callback_id ===
+      SendWelcomeMessageWorkflow.definition.callback_id &&
+    trigger.event_type === "slack#/events/user_joined_channel" &&
+    trigger.channel_ids.includes(channel)
+  ));
+
+  // Return if any matching triggers were found
+  const exists = joinedTriggers.length > 0;
+  return { exists };
+}
+
+/**
+ * saveUserJoinedChannelTrigger creates a new user_joined_channel trigger
+ * for the "Send Welcome Message" workflow in a channel.
+ */
+export async function saveUserJoinedChannelTrigger(
```

Comment:
For my own understanding and lack of having taken this approach first-hand previously: if this is the act of creating a Trigger, why is this here rather than in `triggers/`?

Reply:
A solid question for sure. My reasoning is that this trigger is being created at runtime and requires a dynamic input (the `channel`), while the definitions found in `triggers/` usually seem to be static.

I will create a follow up PR that moves some of this logic into the `triggers/` directory for further discussion since I see a few advantages, disadvantages, and further questions raised by this approach. I think this approach is sufficient for this PR though, so will merge it as is.